File: lingting-spring-web/src/main/java/live/lingting/spring/web/scope/WebScopeHolder.java
package live.lingting.spring.web.scope;

import jakarta.servlet.http.HttpServletRequest;
import live.lingting.framework.util.IpUtils;
import lombok.experimental.UtilityClass;

import java.util.Optional;
import java.util.function.Function;

/**
 * @author lingting 2024-03-20 15:09
 */
@UtilityClass
public class WebScopeHolder {

	static final ThreadLocal<WebScope> LOCAL = new ThreadLocal<>();

	public static WebScope get() {
		return LOCAL.get();
	}

	public static Optional<WebScope> getOptional() {
		return Optional.ofNullable(LOCAL.get());
	}

	public static void set(WebScope webScope) {
		LOCAL.set(webScope);
	}

	public static void remove() {
		LOCAL.remove();
	}

	public static WebScope of(HttpServletRequest request, String traceId, String requestId) {
		return new WebScope(request.getScheme(), request.getHeader("Host"), request.getHeader("Origin"),
				IpUtils.getFirstIp(request), request.getRequestURI(), traceId, requestId,
				request.getHeader("Accept-Language"), request.getHeader("Authorization"),
				request.getHeader("User-Agent"));
	}

	public static <T> T map(Function<WebScope, T> function) {
		return map(function, null);
	}

	public static <T> T map(Function<WebScope, T> function, T other) {
		return getOptional().map(function).orElse(other);
	}

	public static String uri() {
		return map(WebScope::getUri, "");
	}

}

File: lingting-spring-web/src/main/kotlin/live/lingting/spring/web/scope/WebScopeHolder.kt
package live.lingting.spring.web.scope

import jakarta.servlet.http.HttpServletRequest
import live.lingting.framework.context.StackContext
import live.lingting.framework.http.HttpUrlBuilder
import live.lingting.framework.util.IpUtils.getFirstIp
import live.lingting.framework.util.ServletUtils
import java.util.Optional
import java.util.function.Function

/**
 * @author lingting 2024-03-20 15:09
 */
object WebScopeHolder {
    val LOCAL = StackContext<WebScope>()

    @JvmStatic
    fun get(): WebScope? {
        return LOCAL.peek()
    }

    val optional: Optional<WebScope>
        get() = Optional.ofNullable<WebScope>(LOCAL.peek())

    @JvmStatic
    fun put(webScope: WebScope) {
        LOCAL.push(webScope)
    }

    @JvmStatic
    fun pop() {
        LOCAL.pop()
    }

    @JvmStatic
    fun of(request: HttpServletRequest, traceId: String, requestId: String): WebScope {
        val headers = ServletUtils.headers(request)
        val origin = ServletUtils.origin(request, headers)
        var scheme = request.scheme
        var host = headers.hostReal()

        if (!origin.isNullOrBlank()) {
            val builder = HttpUrlBuilder.from(origin)
            scheme = builder.scheme()
            host = builder.headerHost()
        }

        return WebScope(
            scheme,
            host ?: "",
            origin ?: "",
            getFirstIp(request),
            request.requestURI,
            traceId,
            requestId,
            headers.language() ?: "",
            headers.authorization() ?: "",
            headers.ua() ?: "",
            headers
        )
    }

    @JvmStatic
    fun <T> map(function: Function<WebScope, T>): T? {
        return optional.map(function).orElse(null)
    }

    @JvmStatic
    fun <T> map(function: Function<WebScope, T>, other: T): T {
        return optional.map<T>(function).orElse(other)
    }

    @JvmStatic
    fun uri(): String {
        return map<String>(Function { obj -> obj.uri }, "")
    }

}
